python-client/giskard/datasets/base/__init__.py

```diff
@@ -659,16 +659,27 @@ def copy(self):
     def to_mlflow(self, mlflow_client: MlflowClient = None, mlflow_run_id: str = None):
         import mlflow
 
-        with tempfile.NamedTemporaryFile(prefix="dataset-", suffix=".csv") as f:
+        # To avoid file being open in write mode and read at the same time,
+        # First, we'll write it, then make sure to remove it
+        with tempfile.NamedTemporaryFile(
+            prefix="dataset-", suffix=".csv", delete=False
+        ) as f:
+            # Get file path
             local_path = f.name
-            artifact_name = local_path.split("/")[-1]
-            with open(local_path, "wb") as fw:
-                uncompressed_bytes = save_df(self.df)
-                fw.write(uncompressed_bytes)
+            # Get name from file
+            artifact_name = Path(f.name).name
+            # Write the file on disk
+            f.write(save_df(self.df))
+        try:
             if mlflow_client is None and mlflow_run_id is None:
                 mlflow.log_artifact(local_path)
             elif mlflow_client and mlflow_run_id:
                 mlflow_client.log_artifact(mlflow_run_id, local_path=local_path)
+            # Note(Bazire): Should we have a raising else here ?
+        finally:
+            # Force deletion of the temps file
+            Path(f.name).unlink(missing_ok=True)
```

Comment:
deletion from `NamedTemporaryFile`'s `__exit__` didn't do the job?

Reply:
I had to put delete=False when opening tempfile.NamedTemporaryFile

When we are doing 
```
with tempfile.NamedTemporaryFile(
            prefix="dataset-", suffix=".csv", delete=False
        ) as f:
```
It boths create AND open the file (with mode w)

But as stated in the doc (https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile)
>Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows). 

So for this to work for windows, we have to create a temp file, write it, close, read it and then only delete it.

Reply:
ok, sounds good